cudamapper/src/main.cu

```diff
@@ -116,152 +128,206 @@ int main(int argc, char* argv[])
+                                                                                                                                         std::vector<claragenomics::cudamapper::Overlap> filtered_overlaps;
+                                                                                                                                         claragenomics::cudamapper::Overlapper::filter_overlaps(filtered_overlaps, overlaps_to_add);
+                                                                                                                                         overlaps_writer_mtx.lock();
+                                                                                                                                         claragenomics::cudamapper::Overlapper::print_paf(filtered_overlaps);
+                                                                                                                                         overlaps_writer_mtx.unlock();
+                                                                                                                                     },
+                                                                                                                                     overlaps_to_add));
+
+                print_pafs_futures.push_back(write_and_filter_overlaps_future);
             }
         }
+
+        // If all-to-all mapping query will no longer be needed on device, remove it from the cache
+        if (all_to_all)
+        {
+            evict_index(query_start_index, query_end_index, device_id);
+        }
+
+        return print_pafs_futures;
+    };
+
+    // The application (File parsing, index generation, overlap generation etc) is all launched from here.
+    // The main application works as follows:
+    // 1. Launch a worker thread per device (GPU).
+    // 2. Each worker takes target-query ranges off a queue
+    // 3. Each worker pushes vector of futures (since overlap writing is dispatched to an async thread on host). All futures are waited for before the main application exits.
+    std::vector<std::thread> workers;
+    std::atomic<int> ranges_idx(0);
+    std::vector<std::vector<std::shared_ptr<std::future<void>>>> overlap_futures;
+
+    // Launch worker threads
+    for (int device_id = 0; device_id < num_devices; device_id++)
+    {
+        workers.push_back(std::thread(
+            [&, device_id]() {
+                while (ranges_idx < query_target_ranges.size())
+                {
+                    int range_idx = ranges_idx.fetch_add(1);
+                    //Need to perform this check again for thread-safety
+                    if (range_idx < query_target_ranges.size())
+                    {
+                        overlap_futures.push_back(compute_overlaps(query_target_ranges[range_idx], device_id));
```

Comment:
I think this is not thread-safe. Multiple threads are pushing to the same vector.

Reply:
Agreed, resolved